### PR TITLE
[VL] Enrich Spark task OOM error message

### PR DESCRIPTION
**Pull request by zhztheplayer**
The new OOM log message:

```
23/08/24 19:24:08 ERROR ManagedReservationListener: Error reserving memory from target
io.glutenproject.memory.memtarget.ThrowOnOomMemoryTarget$OutOfMemoryException: Not enough spark off-heap execution memory. Acquired: 8388608, granted: 0. Try tweaking config option spark.memory.offHeap.size to get larger space to run this application. 
Current config settings: 
spark.gluten.memory.offHeap.size.in.bytes=1024.0 MiB
spark.gluten.memory.task.offHeap.size.in.bytes=14.2 MiB
Memory consumer stats:
        from GlutenMemoryConsumer/BuildSideRelation#BatchSerializer@2d68d411: Current used bytes: 8.0 MiB, peak bytes: 8.0 MiB
        from                GlutenMemoryConsumer/WholeStageIterator@218fcda2: Current used bytes: 8.0 MiB, peak bytes: 8.0 MiB
        from              GlutenMemoryConsumer/ArrowContextInstance@32393da0: Current used bytes:   0.0 B, peak bytes: 8.0 MiB
        from                                OverAcquire.DummyTarget@5aa0300e: Current used bytes:   0.0 B, peak bytes: 2.4 MiB
	at io.glutenproject.memory.memtarget.ThrowOnOomMemoryTarget.borrow(ThrowOnOomMemoryTarget.java:77)
	at ...
```

### CODE REVIEW

File: gluten-core/src/main/java/io/glutenproject/memory/MemoryUsageStats.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.glutenproject.memory;
+
+public class MemoryUsageStats {
+  public final long peak;
+  public final long current;
+
+  public MemoryUsageStats(MemoryUsage usage) {
```
Comment (Yohahaha) [inline gluten-core/src/main/java/io/glutenproject/memory/MemoryUsageStats.java:23]: Could we pass just long value?

File: gluten-core/src/main/java/io/glutenproject/memory/memtarget/ThrowOnOomMemoryTarget.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+  public ThrowOnOomMemoryTarget(TaskManagedMemoryTarget target) {
+    this.target = target;
+  }
+
+  @Override
+  public long borrow(long size) {
+    long granted = target.borrow(size);
+    if (granted < size) {
+      // OOM
```
Comment (Yohahaha) [inline gluten-core/src/main/java/io/glutenproject/memory/memtarget/ThrowOnOomMemoryTarget.java:34]: Maybe add more description here, "Can not acquire enough memory after spill, throw OOM"

File: gluten-core/src/main/java/io/glutenproject/memory/memtarget/ThrowOnOomMemoryTarget.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+      if (granted != 0L) {
```
Comment (Yohahaha) [inline gluten-core/src/main/java/io/glutenproject/memory/memtarget/ThrowOnOomMemoryTarget.java:35]: this check can be removed, right?

File: gluten-core/src/main/java/io/glutenproject/memory/memtarget/ThrowOnOomMemoryTarget.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                  + "Try tweaking config option spark.memory.offHeap.size to get larger space "
+                  + "to run this application. %n",
+              size, granted));
+      // Log memory usage
+      target.getTaskMemoryManager().showMemoryUsage();
+      // Dump usages to exception body
+      errorBuilder.append(SparkMemoryUtil.dumpMemoryConsumerStats(target.getTaskMemoryManager()));
+      throw new OutOfMemoryException(errorBuilder.toString());
+    }
+    return granted;
```
Comment (Yohahaha) [inline gluten-core/src/main/java/io/glutenproject/memory/memtarget/ThrowOnOomMemoryTarget.java:52]: nit: if (success acquired) { // logic } else { // OOM logic } would more readable.

File: gluten-core/src/main/java/io/glutenproject/memory/MemoryUsageStats.java (hunk 1 above)
Comment (zhztheplayer) [reply gluten-core/src/main/java/io/glutenproject/memory/MemoryUsageStats.java:23]: The class should be tightly coupled with MemoryUsage.java. We can treat it as a snapshot to MemoryUsage instances.

I'll enclose the ctor to package-private.

File: gluten-core/src/main/java/io/glutenproject/memory/memtarget/ThrowOnOomMemoryTarget.java (hunk 3 above)
Comment (zhztheplayer) [reply gluten-core/src/main/java/io/glutenproject/memory/memtarget/ThrowOnOomMemoryTarget.java:35]: `target.repay(0)` may lead to unexpected behavior if the underlying target wasn't coded with respect to this case.

Let's keep the check as some sort of defensive practice.

File: gluten-core/src/main/java/io/glutenproject/memory/memtarget/ThrowOnOomMemoryTarget.java (hunk 4 above)
Comment (zhztheplayer) [reply gluten-core/src/main/java/io/glutenproject/memory/memtarget/ThrowOnOomMemoryTarget.java:52]: Good suggestion however I'd omit the "else" block with fail-fast in if block. Let me make the change.

File: gluten-core/src/main/java/io/glutenproject/memory/memtarget/ThrowOnOomMemoryTarget.java (hunk 2 above)
Comment (zhztheplayer) [reply gluten-core/src/main/java/io/glutenproject/memory/memtarget/ThrowOnOomMemoryTarget.java:34]: Sure, I'll add some comment here.

I'd use

```
// OOM happens.
// Note if the target type is a Spark memory consumer, spilling should already be requested but
// failed to reclaim enough memory. 
```

**Review by zhouyuan — APPROVED** [review]
👍

**Pull request merged**